### ISSUE TITLE
Animation Player FPS shows rounded integer but still internally stores float converted from Seconds

### ISSUE THREAD

**Issue by hsandt** [godotengine/godot#97548]
### Tested versions

- Reproducible in: v4.3.stable.official [77dcf97d8]
- Not reproducible in: v4.2.1.stable.official [b09f793f5]

### System information

Godot v4.3.stable - Ubuntu 22.04.4 LTS 22.04 - X11 - GLES3 (Compatibility) - NVIDIA GeForce GTX 860M (nvidia; 535.183.01) - Intel(R) Core(TM) i7-4710HQ CPU @ 2.50GHz (8 Threads)

### Issue description

**Context**

I am working with fractional FPS for animations, since I set frame time in ms (in Aseprite) then calculate the corresponding FPS for an Animated Sprite (1000/frame duration in ms), then use a converter ([Animated Sprite to Animation Player Convertor for Godot 4.0](https://godotengine.org/asset-library/asset/1605)) from Animated Sprite to Animation Player animation. This leads to FPS such as 12.5 FPS for 80ms frames.

So far in Godot v4.2.1 I could enter either 12.5 FPS or 80ms at the bottom of the Animation Player to get the snapping I wanted:

![image](https://github.com/user-attachments/assets/c1e4a464-a6c9-44e7-b960-ed587d2130bc)

**Issue**

Since Godot 4.3, the FPS field only allows entering an integer. Entering a float will round it and display the rounded integer. This leads to imperfect FPS such as 13 instead of 12.5, leading to snapping between frames (observe the vertical blue bar not being snapped to the start of a sprite square preview):

![image](https://github.com/user-attachments/assets/8760cddd-8cca-450a-8966-6951215607f2)

![image](https://github.com/user-attachments/assets/6fba2cfc-3335-431b-a9c4-733ebf7ed92d)

I need to switch back to Seconds, then enter 0.08 and then switch back to FPS to get 12.5 FPS internally:

![image](https://github.com/user-attachments/assets/a0b76fa5-92ed-402a-ba22-03839b2797f1)

![image](https://github.com/user-attachments/assets/6d57065e-c6f4-47d9-b677-1b4f8171df31)

However, note that the rounded value 13 FPS is still displayed:

![image](https://github.com/user-attachments/assets/701d0aa8-2632-49ff-83c9-61a1910a2307)

although snapping shows we are really still at 12.5 FPS / 80ms:

![image](https://github.com/user-attachments/assets/64b833c1-8017-4cf6-a42a-55d737ac4fac)

This is confusing, and furthermore, trying to re-enter 13 FPS manually will not change the value to 13. However, entering a different value like 14, then the old value 13 again *will* force refresh to 13 instead of 12.5.

**Fix suggestion**

Revert to showing fractional FPS as before, since they are still stored internally as such as indirectly accessibly via setting Seconds, but this requires an extra step for the user.

Currently there is not even an up/down arrow widget to increase/decrease FPS by 1, so there is no big advantage in showing integers anyway (and we could still add an up/down arrow for people who really want to use integers if we want to).

### Steps to reproduce

- Create an Animation Player node with a dummy animation track (the easiest to test timeline snapping is to manually increase the animation duration near the clock icon at the top-right)
- In the Animation Player panel, try to change FPS to a fractional value like 12.5 => rounded to 13
- Move the timeline vertical bar around to test snapping. Add some keyframes there to remember the positions.
- Switch to Seconds, enter 0.08, switch back to FPS => see 13 but internally it's 12.5
- Move the timeline vertical bar around to test snapping. See how it ends at different positions that the previous keyframes.
- Enter 14 FPS, confirm, then 13 again, confirm (to force value refresh).
- Move the timeline vertical bar around to test snapping. See how it's now snapping to the previous keyframes.

### Minimal reproduction project (MRP)

N/A

**Comment by AtlaStar** [comment]
Looks possibly related to #92670
![image](https://github.com/user-attachments/assets/0ee1775a-9149-4f2a-a5f1-132cef11d569)
 
Gonna perform a bisect to be certain, but it also appears that it is intentional. It also seems likely this was the PR that produced the issue given this function given that swapping from seconds to FPS just uses the `1.0/step->get_value()` to determine the new snapping value rather than ensuring that the FPS value is an integer value first, while it does appear to round when swapping from FPS to seconds.
 
![image](https://github.com/user-attachments/assets/aaa726a2-2dfe-4db5-83e1-4b6612dd0180)

**Comment by AtlaStar** [comment]
confirmed that commit b83dc9b from PR #92670 is what introduced this bug.

Also noting that it exists on Windows 10, along with every other platform afaict.

**Comment by AtlaStar** [comment]
I am wondering if a solution would be to allow fractional values, but have the result round to the nearest eighth, or sixteenth when  read from the TSCN. This would implicitly fix the issue with loading 60 fps as 59.9999 FPS as it would round up to 60 FPS and negate the precision errors that would arise, and shouldn't be so precise as to allow issues to arise.

**Comment by TokageItLab** [comment]
As I said there, this is intended, not a bug. If https://github.com/godotengine/godot/pull/97569 does not cause error problems after serialization, then it is fine and this issue can be fixed, but if not, then it needs some other approach, like serializing the FPS directly.

**Comment by AtlaStar** [comment]
To clarify the train of thought here; instead of allowing any and all fractional values for the FPS, instead you clamp the entered value to a value rounded to the nearest 16th. So typing in 15.0541 would clamp it to 15.0625 which is 15 and 1/16. Typing in 12.46 would round to 12.4375 or 12 and 7/16. These values would then be used to save to the tscn.

As to what I believe this report was reporting; it seems to be reporting 2 different "bugs"; the first being the change in behavior which your PR closed as noted above which isn't a bug but a change in intended behavior. The second half and actual bug is that changing between FPS and Seconds doesn't actually change the FPS value, just how it is displayed, as you mentioned.

**Comment by TokageItLab** [comment]
Since KeyEditor only allows integers, so the snap-only fix as in solution #97569 was not sufficient, this is reopened. As I explained in #99013, due to accuracy issues, keys snapped in FPS mode cannot be placed at exact times beyond 10 or 100 seconds.

As long as we allow only integers in FPS, we can display what number of frames it is even if the time placement is not exact, but if we allow fractional FPS for KeyEditor, we cannot restore and display times such as 137.5 or 100.0625 from a placed key.

For precise implementation, I guess it is necessary to have a property on the Animation side for setting FPS apart from the step value, but I am concerned that this would be an expensive implementation for the less common case of fractional FPS.

**Comment by AtlaStar** [comment]
Yeah, my fix ended up being more frail than realized, as such I think it should be reverted for now. Any fix involving supporting fractional FPS should likely revert the PR I originally mentioned after doing some fixups to how things are serialized to any generated .tscn and .tres files.

**Comment by bazsupport** [comment]
I'm having similar snapping issues, but one thing I noticed too is that my default key interval is 0.0333:
![Image](https://github.com/user-attachments/assets/dda2f285-0865-4354-bfb8-7cacb0a51f57)

If I change it, it causes hiccups in frame snaps similar to the example shown in the OG post where it's off by a certain amount. I get around this by cutting and repasting at the correct key frame, but it's annoying and didn't use to do this.

**Comment by TokageItLab** [comment]
@bazsupport In 4.4, the FPS mode is separated and can be avoided by disabling the FPS toggle option.

**Comment by bazsupport** [comment]
@TokageItLab ok thanks, so is 0.0333 the new default? I could have sworn it use to be 0.1 or something

**Comment by TokageItLab** [comment]
@bazsupport It is being worked on in https://github.com/godotengine/godot/pull/98414.

**Comment by bazsupport** [comment]
@TokageItLab you're the best, thanks!
